woocommerce/woocommerce-square
# [Cash App Pay] Add support for splitting payments between a Square Gift Card and Cash App Pay.

### All Submissions:



---

### Changes proposed in this Pull Request:
PR adds support for the split payment between Square Gift Card and Cash App Pay if the gift card has insufficient funds. Currently, we already have split payment support for Square Credit card, this PR extend this support for the Cash App Pay.

> [!NOTE]  
> 1.  Splitting payments is currently possible only on the traditional checkout (shortcode-based). As gift card support has not yet been added to the block checkout (see https://github.com/woocommerce/woocommerce-square/issues/32).
> 2. PR also contains changes from the branch [`enhancement/60`](https://github.com/woocommerce/woocommerce-square/tree/enhancement/60) to accommodate delayed capture. If PR https://github.com/woocommerce/woocommerce-square/pull/62 has not been merged yet, please compare the differences using this link: https://github.com/woocommerce/woocommerce-square/compare/enhancement/60...enhancement/64

> [!IMPORTANT]  
> This PR is not tested with the Square production environment, it should be tested with the Square production environment before we merge/release this.


Closes #64 

### Steps to test the changes in this Pull Request:
1. Add any product to the cart (The product price should be higher than the gift card value).
1. Go to Checkout (shortcode checkout).
1. Enter the gift card number (`7783 3200 0000 0000`) and click Apply. Make sure that you see a note which mentions tha you have to pay the remaining amount with a credit card and Cash App Pay.
1. Make sure the "Cash App Pay" payment gateway is available in the list of available payment gateways to pay the remaining amount of the order.
1. Complete the remaining payment with Cash App Pay and ensure the order is placed successfully.
1. Verify that the payment method value shows both payment methods with the correct amount paid on the order received page, order list page in the admin, and edit order page.
1. Verify that order transaction details (transaction amount and IDs) for both the gift card and Cash App Pay are in the admin order notes.

**Refund:**
1. Perform a full refund on an order placed with split payment between a gift card and Cash App Pay.
1. Perform a partial refund on an order placed with split payment between a gift card and Cash App Pay.
1. Perform a partial refund (custom amount) on an order placed with split payment between a gift card and Cash App Pay.
    - If the refunded amount is more than the amount charged on the gift card, only the charged amount should be refunded to the gift card. The remaining refund amount should be refunded to the cash app pay.
    - If the refunded amount is less than the amount charged on the gift card, the refund should be first done to the gift card.
    - In the case of multiple refunds using custom amounts, the amount should be refunded to the gift card until the entire amount charged on the gift card is refunded.
    - If there are more refunds (custom amount) but the gift card amount is fully refunded, then further refunds should be done to the cash app pay.

**Delayed Capture:**
1. Go to Cash App Pay settings (WooCommerce > Settings > Payments > Cash App Pay (Square)).
1. Update the transaction type to "Authorization".
1. Place an order with split payment between a Square gift card and Cash App Pay (Follow the steps given above).
1. Verify the order status is On-Hold and the payment isn't captured.
1. Go to edit order and click Capture Charge.
1. Verify the order charge gets captured successfully.

### Changelog entry
> Add - Support for splitting payments between a Square Gift Card and Cash App Pay.
